org.spoofax.terms/src/org/spoofax/terms/TermFactory.java

```diff
@@ -115,9 +132,189 @@ public IStrategoTerm annotateTerm(IStrategoTerm term, IStrategoList annotations)
+        StrategoTuple newTerm = new StrategoTuple(subterms, annotations);
+        replaceStrategoTerm(newTerm, replacee);
+        return newTerm;
+    }
+
+    @Override
+    public IStrategoPlaceholder buildPlaceholder(IStrategoTerm template, @Nullable IStrategoTerm replacee, @Nullable IStrategoList annotations) {
+        StrategoPlaceholder newTerm = new StrategoPlaceholder(placeholderConstructor, template, annotations);
+        replaceStrategoTerm(newTerm, replacee);
+        return newTerm;
+    }
+
+    @Override
+    public IStrategoInt buildInt(int value, @Nullable IStrategoTerm replacee, @Nullable IStrategoList annotations) {
+        StrategoInt newTerm = new StrategoInt(value, annotations);
+        replaceStrategoTerm(newTerm, replacee);
+        return newTerm;
+    }
+
+    @Override
+    public IStrategoReal buildReal(double value, @Nullable IStrategoTerm replacee, @Nullable IStrategoList annotations) {
+        StrategoReal newTerm = new StrategoReal(value, annotations);
+        replaceStrategoTerm(newTerm, replacee);
+        return newTerm;
+    }
+
+    @Override
+    public IStrategoString buildString(String value, @Nullable IStrategoTerm replacee, @Nullable IStrategoList annotations) {
+        if(value.length() <= MAX_POOLED_STRING_LENGTH) {
+            synchronized(usedStrings) {
+                value = usedStrings.intern(value);
+            }
+        }
+        StrategoString newTerm = new StrategoString(value, annotations);
+        replaceStrategoTerm(newTerm, replacee);
+        return newTerm;
+    }
+
+    @Override
+    public IStrategoTerm withAnnotations(IStrategoTerm term, @Nullable IStrategoList annotations) {
+        IStrategoList currentAnnos = term.getAnnotations();
+
+        // Cheap checks
+        if (currentAnnos == annotations) {
+            return term;
+        } else if ((annotations == null || annotations.isEmpty()) && currentAnnos.isEmpty()) {
+            return term;
+        }
+
+        // Ordered from most likely to least likely:
+        if (TermUtils.isAppl(term)) {
+            return buildAppl(((IStrategoAppl)term).getConstructor(), term.getAllSubterms(), term, annotations);
+        } else if (TermUtils.isList(term)) {
+            return buildList(term.getAllSubterms(), term, annotations);
+        } else if (TermUtils.isString(term)) {
+            return buildString(((IStrategoString)term).stringValue(), term, annotations);
+        } else if (TermUtils.isInt(term)) {
+            return buildInt(((IStrategoInt)term).intValue(), term, annotations);
+        } else if (TermUtils.isTuple(term)) {
+            return buildTuple(term.getAllSubterms(), term, annotations);
+        } else if (TermUtils.isReal(term)) {
+            return buildReal(((IStrategoReal)term).realValue(), term, annotations);
+        } else if (term instanceof IStrategoPlaceholder) {
+            return buildPlaceholder(((IStrategoPlaceholder)term).getTemplate(), term, annotations);
+        }
```

Comment:
Use a switch on the term type instead. It's faster and you don't have to worry about ordering of most or least likely.

Reply:
`switch instanceof` is not a thing in Java, or am I misunderstanding you? Also, the advantage of this is that it only depends on the semantics of the `is_` methods.

Reply:
Yeah, you misunderstand. By term type, I'm referring to the [`IStrategoTerm#getTermType()`](https://github.com/metaborg/mb-rep/blob/df07f824f3cc3b5986478e1186f6f9adb87c780e/org.spoofax.terms/src/org/spoofax/interpreter/terms/IStrategoTerm.java#L78-L84) method. [Here](https://github.com/metaborg/mb-exec/blob/73bb79dafe4485ddb671b552efc97304cf5b8b2b/org.spoofax.interpreter.core/src/main/java/org/spoofax/interpreter/library/ssl/SSL_mkterm.java#L33-L48)'s an example of the switch.

Reply:
Your comment prompted the `TermType` PR metaborg/mb-rep#22.